pkg/kubelet/util/swap/swap_util.go

```diff
@@ -0,0 +1,109 @@
+package swap
+
+import (
+	"bytes"
+	"fmt"
+	"k8s.io/klog/v2"
+	"k8s.io/mount-utils"
+	"os"
+	sysruntime "runtime"
+	"strings"
+	"sync"
+)
+
+var (
+	tmpfsNoswapOptionSupported        bool
+	tmpfsNoswapOptionAvailabilityOnce sync.Once
+	swapOn                            bool
+	swapOnErr                         error
+	swapOnOnce                        sync.Once
+	rootDir                           string
+)
+
+const TmpfsNoswapOption = "noswap"
+
+// InitializeRootDir gets a path for a base directory, to which we have write permission.
+// This path will be later use for writing temporary files / mount points / etc.
+func InitializeRootDir(path string) {
+	rootDir = path
+}
+
+func IsTmpfsNoswapOptionSupported(mounter mount.Interface) bool {
+	isTmpfsNoswapOptionSupportedHelper := func() bool {
+		if sysruntime.GOOS == "windows" {
+			return false
+		}
+
+		baseMountDir := rootDir
+		if baseMountDir == "" {
+			klog.ErrorS(fmt.Errorf("swap utils root dir is not initialized, falling back to /tmp"), "")
+		}
+
+		mountDir, err := os.MkdirTemp(baseMountDir, "tmpfs-noswap-test-")
+		if err != nil {
+			klog.InfoS("error creating dir to test if tmpfs noswap is enabled. Assuming not supported", "mount path", mountDir, "error", err)
+			return false
+		}
+
+		defer func() {
+			err = os.RemoveAll(mountDir)
+			if err != nil {
+				klog.ErrorS(err, "error removing test tmpfs dir", "mount path", mountDir)
+			}
+		}()
+
+		err = mounter.MountSensitiveWithoutSystemd("tmpfs", mountDir, "tmpfs", []string{TmpfsNoswapOption}, nil)
+		if err != nil {
+			klog.InfoS("error mounting tmpfs with the noswap option. Assuming not supported", "error", err)
+			return false
+		}
+
+		err = mounter.Unmount(mountDir)
+		if err != nil {
+			klog.ErrorS(err, "error unmounting test tmpfs dir", "mount path", mountDir)
+		}
+
+		return true
+	}
+
+	tmpfsNoswapOptionAvailabilityOnce.Do(func() {
+		tmpfsNoswapOptionSupported = isTmpfsNoswapOptionSupportedHelper()
+	})
+
+	return tmpfsNoswapOptionSupported
+}
+
+// IsSwapOn detects whether swap in enabled on the system by inspecting
+// /proc/swaps. If the file does not exist, an os.NotFound error will be returned.
+// If running on windows, swap is assumed to always be false.
+func IsSwapOn() (bool, error) {
+	isSwapOnHelper := func() (bool, error) {
+		if sysruntime.GOOS == "windows" {
+			return false, nil
+		}
+
+		const swapFilePath = "/proc/swaps"
+		swapData, err := os.ReadFile(swapFilePath)
```

Comment:
So, `os.ReadFile()` internally calls `fstat(2)` in order to get the size of the file it's reading. For a lot of procfs and sysfs (kernfs) objects, the size will be `0` anyway. As such, you can avoid this extra call with a combination of `os.Open()` and then reading a page (4 KiB) worth of data in a single go, for example:

```go
f, err := os.Open(file)
if err != nil {
	return err
}
defer f.Close()

data, err := io.ReadAll(io.LimitReader(f, 4096))
if err != nil {
	return err
}
```

Why page worth of data? The amount of data that procfs or sysfs can store and that can be read in one go is 4 KiB, a default page size (for the sake of simplicity, we assume that there isn't any difference between architectures) on most platforms.

This is not something that must need to be changed since this function is run once and during the kubelet start-up. More of a recommendation or advice.

Reply:
Thanks a lot @kwilczynski, makes total sense.
However TBH I prefer to keep this PR focused as much as possible, and I feel changing this behavior is a bit out-of-scope.

I can definitely submit a PR for improving that as a follow-up. Does that sound reasonable?

Reply:
@iholder101, oh no worries. With you there.

Like I said, this is just a "by the way..." kind of comment, for dealing with that sort of pseudo file systems, etc.

Reply:
Cool :)
I'll cc you at the follow-up PR :+1:

Reply:
That looks like a optimization, and is probably better suited to be factored out as a separate function for reusability. :)